bin/reference_data.py

```diff
@@ -1,142 +1,341 @@
 import os
+import ftplib
 import requests
 import argparse
-import ftplib
-import yaml
+import subprocess
 from datetime import datetime
+from cached_property import cached_property
+from egcg_core import util, rest_communication
 from egcg_core.config import cfg
-from egcg_core.app_logging import logging_default as log_cfg
+from egcg_core.app_logging import AppLogger, logging_default
 from config import load_config
 
-log_cfg.add_stdout_handler()
-app_logger = log_cfg.get_logger('genome_downloader')
-
-
-def _now():
-    return datetime.now().strftime('%d/%m/%Y %H:%M:%S')
 
+logging_default.set_log_level(10)
+logging_default.add_stdout_handler()
 
-def _query_ncbi(eutil, db, **query_args):
-    url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/' + eutil + '.fcgi'
-    params = {'db': db, 'retmode': 'JSON', 'version': '2.0'}
-    params.update(query_args)
-    return requests.get(url, params).json()
 
+class DownloadError(Exception):
+    pass
 
-def list_ids(db, search_term, retmax=20):
-    ids = _query_ncbi('esearch', db, term=search_term, retmax=retmax)['esearchresult']
-    if int(ids['count']) > retmax:
-        app_logger.warning('More than %s results found - try a higher retmax, or narrow the search term', retmax)
 
-    app_logger.info('Found %s ids in %s database for %s', ids['count'], db, search_term)
-    return ids['idlist']
+class Downloader(AppLogger):
+    def __init__(self, species, genome_version=None, upload=True, download=True, manual=False):
+        self.species = species
+        self.ftp_species = species.lower().replace(' ', '_')
+        self.species_dir = self.ftp_species[0].upper() + self.ftp_species[1:]
+        self.genome_version = genome_version or self.latest_genome_version()
+        self.upload = upload
+        self.download = download
+        self.manual = manual
 
+        self.data_dir = os.path.join(
+            cfg['reference_data']['base_dir'],
+            self.species_dir,
+            self.genome_version
+        )
+        self.tools = {t: cfg['reference_data'][t] for t in ('picard', 'gatk', 'samtools', 'tabix', 'bgzip', 'bwa')}
+        self.info('Data dir: %s', self.data_dir)
+
+    def run(self):
+        if self.ensembl_release and not self.manual:
+            self.info('Running automated data download and metadata upload')
+            payload = self.prepare_ensembl_metadata()
+            self.finish_metadata(payload)
+
+            if self.download:
+                self.download_data()
+                self.prepare_data()
```

Comment:
`prepare_data` should be run for automatic and manual download.

Reply:
We should either add a `manual_prepare_data method`, or have `prepare_data` fall back to prompting the user for file names